Refactoring of the parallel-coordinates element.

erzel opened the pull request:
* Converted the code into typescript and introduced several cooperating classes to implement the behavior.
* Changed 'schema' to be an object containing all columns (visible and non visible) with a boolean associated with each one to denote whether it is currently displayed. The old format of the schema is
preserved under the deprecated field 'visibleSchema' of 'schema'. The goal is eventually to make everyone use 'schema' and to remove 'visibleSchema'.

erzel commented:
This PR should also take care of b/129365646 (errorneous brush behavior when linear-scale / log-scale axis contains just one point).

wchargin commented:
Thanks for the PR! I’m OOO until next Wednesday and won’t be able to
take a look until then.

erzel commented:
@stephanwlee , @wchargin PTAL.
Thanks!

stephanwlee commented on tensorboard/plugins/hparams/tf_hparams_query_pane/tf-hparams-query-pane.html:
```diff
@@ -322,22 +322,72 @@
+        // set to 'true' after the getExperiment() RPC returns and
+        // the schema is computed.
+        // TODO(erez): We could have set this to a null value initially, but
+        // some consuming elements rely on visibleSchema always being not null.
```
Separately a d.ts file for capturing the schema would be super useful for later refactors and what not.

stephanwlee commented on tensorboard/plugins/hparams/tf_hparams_utils/tf-hparams-utils.html:
```diff
@@ -124,42 +155,131 @@
+        console.assert(isColumnDisplayed(schema, result));
+        return result;
+      }
+      utils.getAbsoluteColumnIndex = getAbsoluteColumnIndex;
```
If I were you, I'd convert this util file to TypeScript and do without nested functions and assignments like this.

stephanwlee commented on tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/tf-hparams-parallel-coords-plot.ts:
```diff
@@ -0,0 +1,1130 @@
+    
+    // Update the DOM.
+    this._parentsSel = this._parentsSel
+      .data(Array.from(visibleColIndices), /*key=*/ (colIndex  => colIndex));
```
Is the key part strictly required? It looks too simple for that.

stephanwlee commented on tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/tf-hparams-parallel-coords-plot.ts:
```diff
@@ -0,0 +1,1130 @@
+    this._parentsSel = this._parentsSel.enter()
+      .append("g")
+      .classed("axis-parent", true)
+      .merge(this._parentsSel)
```
missing semi-colon at the end.

erzel commented:
>I still have not yet completed by first pass. Before I complete it, let me ask this: would it be possible to >break tf-hparams-parallel-coords-plot.ts into few files? I don't think you will need to do tons of type >refactoring. Preferably file per big item (Interaction vs. Axis. vs. Line) would be nice.

I can certainly do that. I don't think it's a big change. I do sometimes find it easier to read through code of related classes when they are all in one file; however, I agree that here this file of ~1000 lines is too big.

erzel commented:
Thanks @stephanwlee! PTAL.

erzel commented:
BTW. I'm stil not a 100% used to github's flow of review, so I may miss some of your comments. If I left any unreplied, please just remind me in your response.

stephanwlee commented on tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/axes.ts:
```diff
@@ -0,0 +1,547 @@
+   * object control which axes are visible.
+   */
+  public updateAxes(options: any, sessionGroups: any[]) {
+    console.assert(!this.isAxisDragging());
```
Instead of throwing exception, I'd silently exit or say something more user friendly especially if it is going to litter in the console.

stephanwlee commented on tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/axes.ts:
```diff
@@ -0,0 +1,547 @@
+
+    // Traverse options.columns, and update each corresponding axis.
+    const visibleColIndices: Set<number> = new Set<number>();
+    options.columns.forEach(c => {
```
for readability, please change `c` to `column`

stephanwlee commented on tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/axes.ts:
```diff
@@ -0,0 +1,547 @@
+    
+    // Update the DOM.
+    this._parentsSel = this._parentsSel
+      .data(Array.from(visibleColIndices), /*key=*/ (colIndex  => colIndex));
```
extraneous whitespace after `colIndex`

stephanwlee commented on tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/axes.ts:
```diff
@@ -0,0 +1,547 @@
+    this._stationaryAxesPositions.domain(
+      /* TypeScript doesn't allow spreading a Set, so we convert to an 
+         Array first. */
+      newDomain.concat(...Array.from(visibleColIndices)));
```
FYI, the spread is extraneous.
```js
> [1].concat([2,3], [4])
< (4) [1, 2, 3, 4]

> [1].concat(...[2,3])
< (3) [1, 2, 3]
```

stephanwlee replied:
Separately, I am a bit confused by Line 508 and this line.

Line 506: `newDomain = visibileColIndices ∩ this._stationaryAxesPositions.domain()`
Line 508:  `visibleColIndices  = visibileColIndices - newDomain`
Line 512: `newDomain = newDomain ∪ visibleColIndices = originalVisibileColIndices`

It might be my afternoon drowsiness but I must be missing something (or can't comprehend code).

erzel commented:
> Could you type the SessionGroup? I see it repeatedly and forget what it should look like.
Done.

erzel commented:
@stephanwlee, friendly ping.

stephanwlee commented on tensorboard/plugins/hparams/tf_hparams_query_pane/tf-hparams-query-pane.html:
```diff
@@ -580,29 +635,66 @@
+        this.schema.metricColumns[metricIndex].displayed =
+            this._metrics[metricIndex].displayed;
+        this._updateVisibleSchema();
+        this.notifyPath(`schema.metricColumns.${metricIndex}.displayed`);
```
I believe if you use `set` API, you don't need to notify like this.
```ts
        this.set(`schema.metricColumns.${metricIndex}.displayed`, this._metrics[metricIndex].displayed);
```

(Also, instead of imperatively calling `updateVisibleSchema`, make sure there is an observer that, on metricColumn display changes, trigger `updateVisibleSchema`.

Similar concern above for `_hparamVisibilityChanged`.

erzel replied:
The issue here is with atomically modifying schema so that observing elements will see the changes to both visibleSchema and metricColumns together. If I use the 'set' API then the notification for the changed metricColumns property will go out immediately, and an observing element will see a schema object in an incosistent state--one whose visibleSchema does not match the new 'metricColumn' information. This may create problems: for example, it would make the assumptions that the code in tf.hparam.utils.getAbsoluteColumnIndex() not hold.
I know this is not ideal, but I don't know of a better solution. I tried to explain that in the comment above the code. I'm open to suggestions.

erzel replied:
> While your rationale makes sense, it looks rather brittle or unconventional. How about this?

I agree. I don't like it very much either, and I'd be happy with a better solution.
Using: 'this.schema = this._computeSchema()' will replace the identity of the schema object each time the user changes the visibility of a single column. This makes it harder for for observing elements to redraw only the changed part. See also the comments in lines 345--350. For example, if the schema object stays the same and we only change its properties, the vaadin-grid element can set the column property  'hidden' to be linked to the (negation of) the corresponding 'displayed' subproperty of schema.hparamColumn or schema.metricColumn.

stephanwlee replied:
If the point of the PR is a refactor, I would advise against such performance optimization (without carefully looking at what is costing the most CPU cycle; i.e., premature optimization). Also, there are many other ways to mitigate such issue while keeping your code sensible. 

One small gotcha with this approach was, which I found while poking around with your branch, with current code, both [1] and [2] while [3] does not (missing `*`; i.e., `schema.*`. Without [3] getting triggered, InteractionManager _can_ go out of sync with the `options` which fortuitously did not happen because we were mutating schema by reference. However, while I could not trigger it, I think they could really go out of sync soon (depending on execution order) when [4] happens. The point here is that, while the bug I pointed out _is_ solvable, it just makes everything more brittle and harder to reason without concerning the greater system. 

[1]: https://github.com/erzel/tensorboard/blob/par_coords_refactor/tensorboard/plugins/hparams/tf_hparams_scale_and_color_controls/tf-hparams-scale-and-color-controls.html#L184
[2]: https://github.com/erzel/tensorboard/blob/par_coords_refactor/tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/tf-hparams-parallel-coords-plot.html#L227-L230
[3]: https://github.com/erzel/tensorboard/blob/c8808d369b1fbb79f270e0053b86becbced82632/tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/tf-hparams-parallel-coords-plot.html#L197
[4]: https://github.com/erzel/tensorboard/blob/c8808d369b1fbb79f270e0053b86becbced82632/tensorboard/plugins/hparams/tf_hparams_query_pane/tf-hparams-query-pane.html#L562

stephanwlee commented on tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/axes.ts:
```diff
@@ -0,0 +1,554 @@
+    visibleColIndices = new Set<number>(visibleColIndices);    
+    let newDomain: number[] = this._stationaryAxesPositions.domain().filter(
+      colIndex => visibleColIndices.has(colIndex));
+    newDomain.forEach(colIndex => visibleColIndices.delete(colIndex));
```
I don't think comments are panacea and making code a little bit readable will help. Please take a look at code below and tell me what you think.
```ts
// Sort visibleColIndices so indicies present in _stationaryAxesPositions
// appear before others.
const visibleDomain = this._stationaryAxesPositions.domain()
     .filter(colIndex => visibleColIndices.has(colIndex));
const newDomain  = Array.from(new Set([
    ...Array.from(visibleDomain),
    ...Array.from(visibleColIndices),
]);
this._stationaryAxesPositions.domain(newDomain);
 ```

stephanwlee commented on tensorboard/plugins/hparams/api.d.ts:
```diff
@@ -0,0 +1,45 @@
+/*
+@license
```
Applicable to other files: not sure what the deal is but other TypeScript file has a bit different style for the license header. Might want to follow those instead of `@license`.
i.e.,
```
/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/
```

stephanwlee commented on tensorboard/plugins/hparams/tf_hparams_parallel_coords_plot/lines.ts:
```diff
@@ -0,0 +1,381 @@
+        () => {
+          const _this = this;
+          this._fgPathsSel.each(
+            function(sessionGroup) {
```
Might work: you can write below to document what `this` means.
```ts
function(this: SVGPathElement, sessionGroup) {
...
}
```

stephanwlee commented on tensorboard/plugins/hparams/tf_hparams_scale_and_color_controls/tf-hparams-scale-and-color-controls.html:
```diff
@@ -251,19 +251,17 @@
+        // Notify polymer that the column changed so that other elements
+        // that depend on it will see the change.
+        const model = event.model;
+        this.notifyPath(`options.columns.${model.column.index}.scale`);
```
Wow, I am very surprised how things are not working properly. I have tried to figure out what is going on underneath to learn quirks about Polymer I may not know about. I tried breaking `options` into smaller private props and tried to make the notifyPath smaller but that didn't help at all. It is super unfortunate that this has to happen but since we, collectively, cannot figure out why, let's just let this be.

erzel replied:
Ok. Thanks for trying to find a better alternative.

erzel commented:
Thanks @stephanwlee! PTAL.

erzel commented:
Hi @stephanwlee,

I've finally managed to incoporate the changes we talked about. I'm sorry for the delay, I'm in a conference this week.
I've split the columns visibility status ('displayed' field) out of 'schema' into a separate boolean array indexed by the column-index. I've done this so that it will be easier for elements to differentiate between the schema changing (e.g. a domain of an hparam) vs a column becoming visible or not.
To make the changes "atomic" I've wrapped all of the relevant information (schema, columnsVisibility and the previous 'visibleSchema') into a single immutable 'configuration object' that is generated by tf-hparams-query-pane. As we discussed, to avoid a race condition between this object and the options object stored in the tf-hparams-scale-and-color-controls, I've added it as a property of options.

PTAL.

Thanks!

stephanwlee commented on tensorboard/plugins/hparams/tf_hparams_query_pane/tf-hparams-query-pane.html:
```diff
@@ -320,72 +320,61 @@
+         *      hparamColumns: Array<{hparamInfo: HParamInfo}>,
+         *      metricColumn: Array<{metricInfo: MetricInfo>}>,
+         *    }
+         *    columnsVisibility: Array<Boolean>,
```
The array here can be very confusing. Instead, may I suggest, maybe in a follow up, use an object where it is keyed by name of the column (assuming that they are unique? -- I cannot imagine it not being unique as it will confuse user too :)) 

Otherwise, this user of this prop has to be aware that hparam columns appear before metric ones. 

Alternative would be, there being flat `columns` array and have a separate map/list of hparam & metrics column ids

stephanwlee commented on tensorboard/plugins/hparams/tf_hparams_query_pane/tf-hparams-query-pane.html:
```diff
@@ -638,48 +631,31 @@
-            this._hparams[hparamIndex].displayed;
-        this._updateVisibleSchema();
-        this.notifyPath(`schema.hparamColumns.${hparamIndex}.displayed`);
+      _hparamVisibilityChanged() {
```
Instead of this, I would recommend using idiomatic Polymer: http://go/gh/erzel/tensorboard/pull/2/files#diff-c777620714a33699c59f411f55c718c4R545 

Also, consider debouncing at least a microtask   :)

stephanwlee commented on tensorboard/plugins/hparams/tf_hparams_session_group_values/tf-hparams-session-group-values.html:
```diff
@@ -67,17 +67,18 @@
-        // in the schema will be displayed. See tf-hparams-query-pane for
-        // the definition of this object.
-        schema: {
+        // The visibleSchema defining the columns to use.
```
super nit: the line wrapping is weird :\

stephanwlee commented on tensorboard/plugins/hparams/tf_hparams_scatter_plot_matrix_plot/tf-hparams-scatter-plot-matrix-plot.html:
```diff
@@ -641,7 +646,9 @@
-          const brushSelection = d3.brushSelection(cellGNode);
+          // For some reason the closure compiler drops the argument when we
+          // write the call below as 'd3.brushSelection(cellGNode)'.
+          const brushSelection = d3["brushSelection"](cellGNode);
```
No action required. As you have indicated, I couldn't find the reason why Closure deletes the argument from this method by inspecting both d3 code and playing around with defining type of this method. I am okay with this hack for time being.

Optional: Btw, although it is technically not in scope of this work, lines 628/639/642 have argument that are not used.

erzel replied:
Thanks! I removed the arguments.
Could you merge the PR, if you don't have any more comments?
